### DISCUSSION TITLE
i need to receive int value to assign to variable in ESP

### DISCUSSION THREAD

**Discussion by daeynasvistas** [ngraziano/LMICPP-Arduino#19] [General]
ok, so when downlink data from NS server to node..

i need to convert value to Base64 (i use Chirpstack) then send it on Port 2 (as example, i can use API to)
I then need to retreive the value in node.

### Example: 
int value of "24156"

Convert to HEX: **24156** -> 0x5E5C
Convert HEX to Base64: 0x5E5C -> **Xlw=**

now send with chirpstack dashboard or API
![image](https://user-images.githubusercontent.com/2634610/153624848-d61649ef-7ffa-48ef-9f19-f861cc7d5e95.png)

Great...

now get that value with great LMICPP-Arduino

```
union ArrayToInteger {
  byte array[2];
  uint16_t integer;
};

if (LMIC.getDataLen()) { 
      PRINT_DEBUG(1, F("Received %d bytes of payload - Receive class C"), LMIC.getDataLen());
      auto data = LMIC.getData();
      if (data) {
        uint8_t port = LMIC.getPort();
        if (port == 2) {                         

                ArrayToInteger converter; //Create a converter

                converter.array[0]=data[1];
                converter.array[1]=data[0];

                  PRINT_DEBUG(1, F("UNION RESULT:")); 
                  PRINT_DEBUG(1, F("Received %02X HEX"), converter.integer);
                  PRINT_DEBUG(1, F("Received %d Value"), converter.integer);
          }
        else {
          //showText(reinterpret_cast<const char *>(data));
        }

      }
    }
```

![image](https://user-images.githubusercontent.com/2634610/153625654-cee86460-6fe4-4696-bf0f-35625b0d44a4.png)

auto data = LMIC.getData(); // i get the my data...

**but is there a better way ?** i dont what to send text, so i need to convert HEX to base64 in first place..
**but i need to receive int value to assign to variable in ESP.**

can i get value "directly from data array?" or the method i use are ok?

many thanks.. :)

**Comment by ngraziano** [comment]
Hi,
For the device side I always do the casting manually to integer like you (i use bit shift but with an union like you do is fine also) and I check data length to be sure.
```cpp
if(LMIC.getPort() ==2 &&  LMIC.getDataLen() >1) {
    auto data = LMIC.getData();
    somevalue = data[0] + data[1] << 8;
}
```

For the server side, I use TTN so I send the data from node-red and have it make the byte buffer and encode it in base64 to send the needed json to the server. I have to always check the order of the byte (big-endian or small-endian)

For the other way (device -> server) I know this library : https://github.com/thesolarnomad/lora-serialization but I have never try it.
